src/MICore/LaunchOptions.cs

```diff
@@ -83,7 +87,106 @@ public PipeLaunchOptions(string pipePath, string pipeArguments, string pipeComma
             this.PipeCommandArguments = pipeCommandArguments;
             this.PipeCwd = pipeCwd;
 
-            this.PipeEnvironment = (pipeEnvironment != null) ? pipeEnvironment.Select(e => new EnvironmentEntry(e)).ToArray() : new EnvironmentEntry[] { };
+            this.PipeEnvironment = new ReadOnlyCollection<EnvironmentEntry>(pipeEnvironment ?? new List<EnvironmentEntry>(0));
+        }
+
+        private static string gdbPathDefault = @"/usr/bin/gdb";
+        static internal PipeLaunchOptions CreateFromJson(JObject parsedOptions)
+        {
+            Debug.Assert(parsedOptions["pipeTransport"] != null && parsedOptions["pipeTransport"].HasValues, "PipeTransport should exist and have values.");
+
+            Json.LaunchOptions.PipeTransport pipeTransport = parsedOptions["pipeTransport"].ToObject<Json.LaunchOptions.PipeTransport>();
+
+            // PipeProgram must be specified
+            if (String.IsNullOrWhiteSpace(pipeTransport.PipeProgram))
+            {
+                throw new InvalidLaunchOptionsException(String.Format(CultureInfo.InvariantCulture, MICoreResources.Error_EmptyPipePath));
+            }
+
+            string pipeProgram = EnsurePipeProgramExists(pipeTransport);
```

Comment:
This function doesn't handle the case that pipeProgram is a command on the PATH. You might want to omit this and just deal with Process.Start failing, which seems to be what we do for XML.